core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -547,19 +536,30 @@ class TransactionStateManager(brokerId: Int,
             val metadata = epochAndTxnMetadata.transactionMetadata
             metadata synchronized {
               if (epochAndTxnMetadata.coordinatorEpoch == coordinatorEpoch) {
-                debug(s"TransactionalId ${metadata.transactionalId}, resetting pending state since we are returning error $responseError")
-                metadata.pendingState = None
+                if (retryOnError(responseError)) {
```

Comment:
Hmm.. I see several new unit tests which mock `appendTransactionToLog`, but none which test it itself.